{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "332b80b7-952d-4edb-b411-8693c8277273",
   "metadata": {},
   "source": [
    "# RAxML-NG on 50% ID MSA\n",
    "\n",
    "Aim: Use RAxML-NG to generate a phylogenetic tree for my protein. \n",
    "This notebook documents the process I used:\n",
    "1. Building the Maximum-liklhood (ML) tree\n",
    "2. Generating bootstrap replicates\n",
    "3. Mapping bootsrap support onto the ML tree\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "404f3d74-d5ea-4dab-99ad-b06b144f856b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MSA: /Users/lachlanblack/Documents/GitHub/EspY3-protein-evolution/data/cleaned/trimal/espy3-50.mafft.r07s08.trim.fasta\n",
      "OUTDIR: /Users/lachlanblack/Documents/GitHub/EspY3-protein-evolution/data/cleaned/raxml/test/01_50id_bootstrap\n",
      "MODEL: Blosum62+G4\n"
     ]
    }
   ],
   "source": [
    "from pathlib import Path # allows handling of files and folder paths cleanly\n",
    "import subprocess # allows terminal commands to be run from python\n",
    "\n",
    "# PARAMETERS\n",
    "MSA = Path(\"../../data/cleaned/trimal/espy3-50.mafft.r07s08.trim.fasta\") # trimmed MSA\n",
    "OUTDIR = Path(\"../../data/cleaned/raxml/test/01_50id_bootstrap\") # output folder for RaxML results\n",
    "PREFIX = \"mytree\" # label to prefix all the output files\n",
    "MODEL = \"Blosum62+G4\" # substitution model used for protein data\n",
    "BS_TREES = 100 # number of bootstrap replicates\n",
    "THREADS = \"auto\" # use all available CPU threads automatically\n",
    "\n",
    "# Make sure output folder exists\n",
    "OUTDIR.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "# Print settings to confirm they are correct\n",
    "print (f\"MSA: {MSA.resolve()}\")\n",
    "print(f\"OUTDIR: {OUTDIR.resolve()}\")\n",
    "print(f\"MODEL: {MODEL}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3b810b75-fd2a-4adf-8fb7-95cc452c7599",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define small helper function to run terminal commands\n",
    "def run(cmd):\n",
    "    print(f\"\\n$ {cmd}\\n\") # prints command to show exactly whas ran\n",
    "    process = subprocess.run(cmd, shell=True) # runs command in terminal\n",
    "    if process.returncode != 0: # checks if it failed (non-zero = error)\n",
    "        raise RuntimeError(f\"Command failed: {cmd}\") # stop notebook if it failed\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "704a1489-4c00-48d6-a5c2-fb1efe2fa3cc",
   "metadata": {},
   "source": [
    "Explanation of the RAxML-NG command:\n",
    "- `--search` -> perform a maximum-likelihood search\n",
    "- `--msa` -> the MSA file used to make tree\n",
    "- `--model` -> the substitution model (Blosum62+G4)\n",
    "- `--prefix` -> sets output file prefix\n",
    "- `--threads auto` -> use all available CPU cores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cdce01e4-794c-41eb-b39d-ab5be2cdd662",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "$ raxml-ng --search --msa ../../data/cleaned/trimal/espy3-50.mafft.r07s08.trim.fasta --model Blosum62+G4 --prefix ../../data/cleaned/raxml/test/01_50id_bootstrap/mytree --threads auto --redo\n",
      "\n",
      "\n",
      "RAxML-NG v. 1.2.2 released on 30.04.2024 by The Exelixis Lab.\n",
      "Developed by: Alexey M. Kozlov and Alexandros Stamatakis.\n",
      "Contributors: Diego Darriba, Tomas Flouri, Benoit Morel, Sarah Lutteropp, Ben Bettisworth, Julia Haag, Anastasis Togkousidis.\n",
      "Latest version: https://github.com/amkozlov/raxml-ng\n",
      "Questions/problems/suggestions? Please visit: https://groups.google.com/forum/#!forum/raxml\n",
      "\n",
      "System: Apple M2, 8 cores, 8 GB RAM\n",
      "\n",
      "RAxML-NG was called at 03-Nov-2025 18:56:17 as follows:\n",
      "\n",
      "raxml-ng --search --msa ../../data/cleaned/trimal/espy3-50.mafft.r07s08.trim.fasta --model Blosum62+G4 --prefix ../../data/cleaned/raxml/test/01_50id_bootstrap/mytree --threads auto --redo\n",
      "\n",
      "Analysis options:\n",
      "  run mode: ML tree search\n",
      "  start tree(s): random (10) + parsimony (10)\n",
      "  random seed: 1762196177\n",
      "  tip-inner: OFF\n",
      "  pattern compression: ON\n",
      "  per-rate scalers: OFF\n",
      "  site repeats: ON\n",
      "  logLH epsilon: general: 10.000000, brlen-triplet: 1000.000000\n",
      "  fast spr radius: AUTO\n",
      "  spr subtree cutoff: 1.000000\n",
      "  fast CLV updates: ON\n",
      "  branch lengths: proportional (ML estimate, algorithm: NR-FAST)\n",
      "  SIMD kernels: SSE3\n",
      "  parallelization: coarse-grained (auto), PTHREADS (auto)\n",
      "\n",
      "WARNING: Running in REDO mode: existing checkpoints are ignored, and all result files will be overwritten!\n",
      "\n",
      "[00:00:00] Reading alignment from file: ../../data/cleaned/trimal/espy3-50.mafft.r07s08.trim.fasta\n",
      "[00:00:00] Loaded alignment with 44 taxa and 524 sites\n",
      "\n",
      "WARNING: Sequences tr|A0A0D8WGH1|A0A0D8WGH1_ECOLX/1-523 and tr|A0A377DGX1|A0A377DGX1_ECOLX/1-523 are exactly identical!\n",
      "WARNING: Sequences tr|A0A1Q5ZEH3|A0A1Q5ZEH3_ECOLX/1-523 and tr|A0A3R8KSB7|A0A3R8KSB7_ECOLX/1-523 are exactly identical!\n",
      "WARNING: Sequences tr|A0AAW5Z1M4|A0AAW5Z1M4_ECOLX/1-523 and tr|A0AAW6W494|A0AAW6W494_ECOLX/1-523 are exactly identical!\n",
      "WARNING: Sequences tr|A0A0X5K2I6|A0A0X5K2I6_ECOLX/1-523 and tr|W1X2U4|W1X2U4_ECOLX/1-523 are exactly identical!\n",
      "WARNING: Sequences tr|A0A0X5K2I6|A0A0X5K2I6_ECOLX/1-523 and tr|A0A376VQ42|A0A376VQ42_ECOLX/1-523 are exactly identical!\n",
      "WARNING: Sequences tr|A0A0X5K2I6|A0A0X5K2I6_ECOLX/1-523 and tr|W1XNK9|W1XNK9_9ZZZZ/1-523 are exactly identical!\n",
      "WARNING: Sequences tr|A0A0X5K2I6|A0A0X5K2I6_ECOLX/1-523 and tr|A0A6D1PLN0|A0A6D1PLN0_ECOLX/1-523 are exactly identical!\n",
      "WARNING: Sequences tr|A0A0X5K2I6|A0A0X5K2I6_ECOLX/1-523 and tr|A0A8S7N1T2|A0A8S7N1T2_ECOLX/1-523 are exactly identical!\n",
      "WARNING: Sequences tr|A0A0X5K2I6|A0A0X5K2I6_ECOLX/1-523 and tr|A0AAW5ZB89|A0AAW5ZB89_ECOLX/1-523 are exactly identical!\n",
      "WARNING: Sequences tr|W1YEN7|W1YEN7_9ZZZZ/1-523 and tr|A0A0H3PZQ4|A0A0H3PZQ4_ECO5C/1-523 are exactly identical!\n",
      "WARNING: Sequences tr|A0A767FIQ4|A0A767FIQ4_ECOLX/1-520 and tr|A0A793GC28|A0A793GC28_ECOLX/1-520 are exactly identical!\n",
      "WARNING: Sequences tr|A0A767FIQ4|A0A767FIQ4_ECOLX/1-520 and tr|A0A9P2PS47|A0A9P2PS47_ECOLX/1-520 are exactly identical!\n",
      "WARNING: Sequences tr|A0A767FIQ4|A0A767FIQ4_ECOLX/1-520 and tr|A0A9Q6ZFC8|A0A9Q6ZFC8_ECO57/1-520 are exactly identical!\n",
      "WARNING: Sequences tr|Q8XE85|Q8XE85_ECO57/1-518 and tr|A0A793GH32|A0A793GH32_ECOLX/1-518 are exactly identical!\n",
      "WARNING: Duplicate sequences found: 14\n",
      "\n",
      "NOTE: Reduced alignment (with duplicates and gap-only sites/taxa removed) \n",
      "NOTE: was saved to: /Users/lachlanblack/Documents/GitHub/EspY3-protein-evolution/data/cleaned/raxml/test/01_50id_bootstrap/mytree.raxml.reduced.phy\n",
      "\n",
      "Alignment comprises 1 partitions and 185 patterns\n",
      "\n",
      "Partition 0: noname\n",
      "Model: BLOSUM62+G4m\n",
      "Alignment sites / patterns: 524 / 185\n",
      "Gaps: 0.61 %\n",
      "Invariant sites: 71.37 %\n",
      "\n",
      "\n",
      "NOTE: Binary MSA file created: ../../data/cleaned/raxml/test/01_50id_bootstrap/mytree.raxml.rba\n",
      "\n",
      "Parallelization scheme autoconfig: 8 worker(s) x 1 thread(s)\n",
      "\n",
      "[00:00:00] Generating 10 random starting tree(s) with 44 taxa\n",
      "[00:00:00] Generating 10 parsimony starting tree(s) with 44 taxa\n",
      "Parallel parsimony with 8 threads\n",
      "Parallel reduction/worker buffer size: 1 KB  / 0 KB\n",
      "\n",
      "[00:00:00] Data distribution: max. partitions/sites/weight per thread: 1 / 185 / 14800\n",
      "[00:00:00] Data distribution: max. searches per worker: 3\n",
      "\n",
      "Starting ML tree search with 20 distinct starting trees\n",
      "\n",
      "[00:00:00 -7779.856274] Initial branch length optimization\n",
      "[00:00:00 -6520.859904] Model parameter optimization (eps = 10.000000)\n",
      "[00:00:00 -6019.225742] AUTODETECT spr round 1 (radius: 5)\n",
      "[00:00:01 -4448.766652] AUTODETECT spr round 2 (radius: 10)\n",
      "[00:00:03 -3357.142949] AUTODETECT spr round 3 (radius: 15)\n",
      "[00:00:04 -3321.955631] AUTODETECT spr round 4 (radius: 20)\n",
      "[00:00:04 -3321.805704] AUTODETECT spr round 5 (radius: 25)\n",
      "[00:00:04 -3321.723257] SPR radius for FAST iterations: 20 (autodetect)\n",
      "[00:00:04 -3321.723257] Model parameter optimization (eps = 3.000000)\n",
      "[00:00:04 -3310.945730] FAST spr round 1 (radius: 20)\n",
      "[00:00:06 -3204.585257] FAST spr round 2 (radius: 20)\n",
      "[00:00:07 -3202.121030] Model parameter optimization (eps = 1.000000)\n",
      "[00:00:07 -3201.385613] SLOW spr round 1 (radius: 5)\n",
      "[00:00:10 -3201.145952] SLOW spr round 2 (radius: 10)\n",
      "[00:00:12 -3201.145910] SLOW spr round 3 (radius: 15)\n",
      "[00:00:13] [worker #2] ML tree search #3, logLikelihood: -3201.866835\n",
      "[00:00:13 -3201.145910] SLOW spr round 4 (radius: 20)\n",
      "[00:00:13] [worker #7] ML tree search #8, logLikelihood: -3204.436903\n",
      "[00:00:14] [worker #5] ML tree search #6, logLikelihood: -3202.048142\n",
      "[00:00:14] [worker #3] ML tree search #4, logLikelihood: -3202.045419\n",
      "[00:00:14] [worker #6] ML tree search #7, logLikelihood: -3202.047981\n",
      "[00:00:14 -3201.145910] SLOW spr round 5 (radius: 25)\n",
      "[00:00:14 -3201.145910] Model parameter optimization (eps = 0.100000)\n",
      "\n",
      "[00:00:14] [worker #0] ML tree search #1, logLikelihood: -3201.143411\n",
      "\n",
      "[00:00:14 -7820.070804] Initial branch length optimization\n",
      "[00:00:14 -6696.520755] Model parameter optimization (eps = 10.000000)\n",
      "[00:00:15] [worker #1] ML tree search #2, logLikelihood: -3202.045811\n",
      "[00:00:15 -6202.584920] AUTODETECT spr round 1 (radius: 5)\n",
      "[00:00:15] [worker #4] ML tree search #5, logLikelihood: -3202.047987\n",
      "[00:00:16 -4206.724711] AUTODETECT spr round 2 (radius: 10)\n",
      "[00:00:17 -3483.322225] AUTODETECT spr round 3 (radius: 15)\n",
      "[00:00:18 -3481.189669] AUTODETECT spr round 4 (radius: 20)\n",
      "[00:00:19 -3440.901498] AUTODETECT spr round 5 (radius: 25)\n",
      "[00:00:19 -3440.759620] SPR radius for FAST iterations: 25 (autodetect)\n",
      "[00:00:19 -3440.759620] Model parameter optimization (eps = 3.000000)\n",
      "[00:00:19 -3430.332990] FAST spr round 1 (radius: 25)\n",
      "[00:00:21 -3225.952605] FAST spr round 2 (radius: 25)\n",
      "[00:00:22 -3205.173766] FAST spr round 3 (radius: 25)\n",
      "[00:00:23 -3203.683315] Model parameter optimization (eps = 1.000000)\n",
      "[00:00:23] [worker #2] ML tree search #11, logLikelihood: -3202.047977\n",
      "[00:00:23 -3202.517032] SLOW spr round 1 (radius: 5)\n",
      "[00:00:24] [worker #5] ML tree search #14, logLikelihood: -3202.047980\n",
      "[00:00:24] [worker #6] ML tree search #15, logLikelihood: -3202.048099\n",
      "[00:00:24] [worker #7] ML tree search #16, logLikelihood: -3202.047978\n",
      "[00:00:24] [worker #3] ML tree search #12, logLikelihood: -3204.436903\n",
      "[00:00:25] [worker #4] ML tree search #13, logLikelihood: -3206.226122\n",
      "[00:00:26 -3202.045894] SLOW spr round 2 (radius: 10)\n",
      "[00:00:28 -3202.045845] SLOW spr round 3 (radius: 15)\n",
      "[00:00:29] [worker #1] ML tree search #10, logLikelihood: -3204.197287\n",
      "[00:00:29 -3202.045845] SLOW spr round 4 (radius: 20)\n",
      "[00:00:30 -3202.045845] SLOW spr round 5 (radius: 25)\n",
      "[00:00:30 -3202.045845] Model parameter optimization (eps = 0.100000)\n",
      "\n",
      "[00:00:30] [worker #0] ML tree search #9, logLikelihood: -3202.045808\n",
      "\n",
      "[00:00:30 -4420.515425] Initial branch length optimization\n",
      "[00:00:31 -3232.459964] Model parameter optimization (eps = 10.000000)\n",
      "[00:00:31 -3220.175556] AUTODETECT spr round 1 (radius: 5)\n",
      "[00:00:32 -3209.176843] AUTODETECT spr round 2 (radius: 10)\n",
      "[00:00:34 -3205.959745] AUTODETECT spr round 3 (radius: 15)\n",
      "[00:00:34] [worker #2] ML tree search #19, logLikelihood: -3202.048143\n",
      "[00:00:34 -3205.618300] AUTODETECT spr round 4 (radius: 20)\n",
      "[00:00:34 -3205.616470] SPR radius for FAST iterations: 15 (autodetect)\n",
      "[00:00:34 -3205.616470] Model parameter optimization (eps = 3.000000)\n",
      "[00:00:34 -3205.590137] FAST spr round 1 (radius: 15)\n",
      "[00:00:35 -3202.048993] Model parameter optimization (eps = 1.000000)\n",
      "[00:00:35 -3202.047982] SLOW spr round 1 (radius: 5)\n",
      "[00:00:35] [worker #3] ML tree search #20, logLikelihood: -3202.047989\n",
      "[00:00:37 -3202.047981] SLOW spr round 2 (radius: 10)\n",
      "[00:00:38] [worker #1] ML tree search #18, logLikelihood: -3202.047978\n",
      "[00:00:39 -3202.047981] SLOW spr round 3 (radius: 15)\n",
      "[00:00:40 -3202.047981] SLOW spr round 4 (radius: 20)\n",
      "[00:00:41 -3202.047981] SLOW spr round 5 (radius: 25)\n",
      "[00:00:41 -3202.047981] Model parameter optimization (eps = 0.100000)\n",
      "\n",
      "[00:00:41] [worker #0] ML tree search #17, logLikelihood: -3202.047981\n",
      "\n",
      "\n",
      "Optimized model parameters:\n",
      "\n",
      "   Partition 0: noname\n",
      "   Rate heterogeneity: GAMMA (4 cats, mean),  alpha: 0.314105 (ML),  weights&rates: (0.250000,0.006509) (0.250000,0.115503) (0.250000,0.618748) (0.250000,3.259240) \n",
      "   Base frequencies (model): 0.074000 0.052000 0.045000 0.054000 0.025000 0.034000 0.054000 0.074000 0.026000 0.068000 0.099000 0.058000 0.025000 0.047000 0.039000 0.057000 0.051000 0.013000 0.032000 0.073000 \n",
      "   Substitution rates (model): 0.735790 0.485391 0.543162 1.459995 1.199706 1.170949 1.955884 0.716241 0.605899 0.800017 1.295201 1.253758 0.492965 1.173276 4.325093 1.729178 0.465839 0.718207 2.187775 1.297447 0.500964 0.227827 3.020834 1.360574 0.418763 1.456141 0.232036 0.622712 5.411115 0.983693 0.371645 0.448134 1.122783 0.914666 0.426382 0.720517 0.438388 3.180100 0.397359 1.839216 1.240489 1.355872 2.414501 0.283017 0.211888 1.593137 0.648441 0.354861 0.494887 2.904102 1.898174 0.191482 0.538223 0.312859 0.240837 1.190946 3.761625 0.798473 0.778143 0.418556 0.218132 1.032448 0.222622 0.281731 0.730628 1.582754 0.934188 0.145345 0.261422 0.258129 0.329802 0.140749 0.418203 0.354058 0.774894 0.831843 0.285079 0.767689 0.441337 0.356008 1.197188 1.119831 0.527664 0.470238 1.116352 5.528919 0.609846 2.435341 0.236202 0.580737 3.945278 2.494896 0.144357 0.858571 1.934871 1.277480 0.758654 0.958990 0.530786 0.423580 1.626891 0.186848 0.372625 2.802427 0.555415 0.291409 0.926564 1.769893 1.071097 0.407636 0.596719 0.524254 0.539859 0.189296 0.217721 0.752042 0.459436 0.368166 0.504087 1.509326 0.641436 0.508359 0.308056 0.253341 0.252718 0.348072 1.022507 0.984312 0.714534 0.527007 1.117030 0.585407 0.301249 4.218954 0.201556 3.890964 0.406194 3.364798 1.517359 0.388355 0.357544 1.179091 0.341986 0.674617 8.311839 0.445570 6.030559 2.064840 0.374556 0.352969 0.915260 0.691475 0.811246 2.231406 1.073061 0.266925 1.047383 1.752166 1.303875 0.332243 0.717993 0.498138 1.773855 0.454124 0.918723 1.488548 0.888101 0.951682 2.575851 0.233598 0.540028 0.488206 2.074325 6.747260 0.838120 1.169130 1.005452 0.252215 0.369405 0.496908 5.151556 0.387926 0.796752 0.561925 0.513128 0.801010 2.253074 4.054419 0.266509 1.000000 \n",
      "\n",
      "\n",
      "Final LogLikelihood: -3201.143411\n",
      "\n",
      "AIC score: 6574.286822 / AICc score: 6608.529385 / BIC score: 6940.775107\n",
      "Free parameters (model + branch lengths): 86\n",
      "\n",
      "WARNING: Best ML tree contains 18 near-zero branches!\n",
      "\n",
      "Best ML tree with collapsed near-zero branches saved to: /Users/lachlanblack/Documents/GitHub/EspY3-protein-evolution/data/cleaned/raxml/test/01_50id_bootstrap/mytree.raxml.bestTreeCollapsed\n",
      "Best ML tree saved to: /Users/lachlanblack/Documents/GitHub/EspY3-protein-evolution/data/cleaned/raxml/test/01_50id_bootstrap/mytree.raxml.bestTree\n",
      "All ML trees saved to: /Users/lachlanblack/Documents/GitHub/EspY3-protein-evolution/data/cleaned/raxml/test/01_50id_bootstrap/mytree.raxml.mlTrees\n",
      "Optimized model saved to: /Users/lachlanblack/Documents/GitHub/EspY3-protein-evolution/data/cleaned/raxml/test/01_50id_bootstrap/mytree.raxml.bestModel\n",
      "\n",
      "Execution log saved to: /Users/lachlanblack/Documents/GitHub/EspY3-protein-evolution/data/cleaned/raxml/test/01_50id_bootstrap/mytree.raxml.log\n",
      "\n",
      "Analysis started: 03-Nov-2025 18:56:17 / finished: 03-Nov-2025 18:56:58\n",
      "\n",
      "Elapsed time: 41.239 seconds\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Build the ML tree\n",
    "\n",
    "# Outputs:\n",
    "    # {PREFIX}.raxml.bestTREE -> Best ML tree\n",
    "    # {PREFIX}.raxml.log -> Log file\n",
    "    # {PREFIX}.raxml.mlTrees -> ML search trees\n",
    "run(f\"raxml-ng --search --msa {MSA} --model {MODEL} --prefix {OUTDIR / PREFIX} --threads {THREADS} --redo\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "74e441de-4682-4a0b-8c5d-6bd1ed4f3baf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "$ raxml-ng --bootstrap --msa ../../data/cleaned/trimal/espy3-50.mafft.r07s08.trim.fasta --model Blosum62+G4 --bs-trees 100 --prefix ../../data/cleaned/raxml/test/01_50id_bootstrap/mytree_bs --threads auto --redo\n",
      "\n",
      "\n",
      "RAxML-NG v. 1.2.2 released on 30.04.2024 by The Exelixis Lab.\n",
      "Developed by: Alexey M. Kozlov and Alexandros Stamatakis.\n",
      "Contributors: Diego Darriba, Tomas Flouri, Benoit Morel, Sarah Lutteropp, Ben Bettisworth, Julia Haag, Anastasis Togkousidis.\n",
      "Latest version: https://github.com/amkozlov/raxml-ng\n",
      "Questions/problems/suggestions? Please visit: https://groups.google.com/forum/#!forum/raxml\n",
      "\n",
      "System: Apple M2, 8 cores, 8 GB RAM\n",
      "\n",
      "RAxML-NG was called at 03-Nov-2025 18:57:12 as follows:\n",
      "\n",
      "raxml-ng --bootstrap --msa ../../data/cleaned/trimal/espy3-50.mafft.r07s08.trim.fasta --model Blosum62+G4 --bs-trees 100 --prefix ../../data/cleaned/raxml/test/01_50id_bootstrap/mytree_bs --threads auto --redo\n",
      "\n",
      "Analysis options:\n",
      "  run mode: Bootstrapping\n",
      "  start tree(s): \n",
      "  bootstrap replicates: parsimony (100)\n",
      "  random seed: 1762196232\n",
      "  tip-inner: OFF\n",
      "  pattern compression: ON\n",
      "  per-rate scalers: OFF\n",
      "  site repeats: ON\n",
      "  logLH epsilon: general: 0.100000, brlen-triplet: 1000.000000\n",
      "  fast spr radius: AUTO\n",
      "  spr subtree cutoff: 1.000000\n",
      "  fast CLV updates: ON\n",
      "  branch lengths: proportional (ML estimate, algorithm: NR-FAST)\n",
      "  SIMD kernels: SSE3\n",
      "  parallelization: coarse-grained (auto), PTHREADS (auto)\n",
      "\n",
      "WARNING: Running in REDO mode: existing checkpoints are ignored, and all result files will be overwritten!\n",
      "\n",
      "[00:00:00] Reading alignment from file: ../../data/cleaned/trimal/espy3-50.mafft.r07s08.trim.fasta\n",
      "[00:00:00] Loaded alignment with 44 taxa and 524 sites\n",
      "\n",
      "WARNING: Sequences tr|A0A0D8WGH1|A0A0D8WGH1_ECOLX/1-523 and tr|A0A377DGX1|A0A377DGX1_ECOLX/1-523 are exactly identical!\n",
      "WARNING: Sequences tr|A0A1Q5ZEH3|A0A1Q5ZEH3_ECOLX/1-523 and tr|A0A3R8KSB7|A0A3R8KSB7_ECOLX/1-523 are exactly identical!\n",
      "WARNING: Sequences tr|A0AAW5Z1M4|A0AAW5Z1M4_ECOLX/1-523 and tr|A0AAW6W494|A0AAW6W494_ECOLX/1-523 are exactly identical!\n",
      "WARNING: Sequences tr|A0A0X5K2I6|A0A0X5K2I6_ECOLX/1-523 and tr|W1X2U4|W1X2U4_ECOLX/1-523 are exactly identical!\n",
      "WARNING: Sequences tr|A0A0X5K2I6|A0A0X5K2I6_ECOLX/1-523 and tr|A0A376VQ42|A0A376VQ42_ECOLX/1-523 are exactly identical!\n",
      "WARNING: Sequences tr|A0A0X5K2I6|A0A0X5K2I6_ECOLX/1-523 and tr|W1XNK9|W1XNK9_9ZZZZ/1-523 are exactly identical!\n",
      "WARNING: Sequences tr|A0A0X5K2I6|A0A0X5K2I6_ECOLX/1-523 and tr|A0A6D1PLN0|A0A6D1PLN0_ECOLX/1-523 are exactly identical!\n",
      "WARNING: Sequences tr|A0A0X5K2I6|A0A0X5K2I6_ECOLX/1-523 and tr|A0A8S7N1T2|A0A8S7N1T2_ECOLX/1-523 are exactly identical!\n",
      "WARNING: Sequences tr|A0A0X5K2I6|A0A0X5K2I6_ECOLX/1-523 and tr|A0AAW5ZB89|A0AAW5ZB89_ECOLX/1-523 are exactly identical!\n",
      "WARNING: Sequences tr|W1YEN7|W1YEN7_9ZZZZ/1-523 and tr|A0A0H3PZQ4|A0A0H3PZQ4_ECO5C/1-523 are exactly identical!\n",
      "WARNING: Sequences tr|A0A767FIQ4|A0A767FIQ4_ECOLX/1-520 and tr|A0A793GC28|A0A793GC28_ECOLX/1-520 are exactly identical!\n",
      "WARNING: Sequences tr|A0A767FIQ4|A0A767FIQ4_ECOLX/1-520 and tr|A0A9P2PS47|A0A9P2PS47_ECOLX/1-520 are exactly identical!\n",
      "WARNING: Sequences tr|A0A767FIQ4|A0A767FIQ4_ECOLX/1-520 and tr|A0A9Q6ZFC8|A0A9Q6ZFC8_ECO57/1-520 are exactly identical!\n",
      "WARNING: Sequences tr|Q8XE85|Q8XE85_ECO57/1-518 and tr|A0A793GH32|A0A793GH32_ECOLX/1-518 are exactly identical!\n",
      "WARNING: Duplicate sequences found: 14\n",
      "\n",
      "NOTE: Reduced alignment (with duplicates and gap-only sites/taxa removed) \n",
      "NOTE: was saved to: /Users/lachlanblack/Documents/GitHub/EspY3-protein-evolution/data/cleaned/raxml/test/01_50id_bootstrap/mytree_bs.raxml.reduced.phy\n",
      "\n",
      "Alignment comprises 1 partitions and 185 patterns\n",
      "\n",
      "Partition 0: noname\n",
      "Model: BLOSUM62+G4m\n",
      "Alignment sites / patterns: 524 / 185\n",
      "Gaps: 0.61 %\n",
      "Invariant sites: 71.37 %\n",
      "\n",
      "\n",
      "NOTE: Binary MSA file created: ../../data/cleaned/raxml/test/01_50id_bootstrap/mytree_bs.raxml.rba\n",
      "\n",
      "Parallelization scheme autoconfig: 8 worker(s) x 1 thread(s)\n",
      "\n",
      "Parallel reduction/worker buffer size: 1 KB  / 0 KB\n",
      "\n",
      "[00:00:00] Data distribution: max. partitions/sites/weight per thread: 1 / 185 / 14800\n",
      "[00:00:00] Data distribution: max. searches per worker: 13\n",
      "[00:00:00] Starting bootstrapping analysis with 100 replicates.\n",
      "\n",
      "[00:00:08] [worker #2] Bootstrap tree #3, logLikelihood: -3220.837662\n",
      "[00:00:08] [worker #4] Bootstrap tree #5, logLikelihood: -3324.230722\n",
      "[00:00:08] [worker #6] Bootstrap tree #7, logLikelihood: -3159.522789\n",
      "[00:00:08] [worker #5] Bootstrap tree #6, logLikelihood: -3158.544708\n",
      "[00:00:08] [worker #1] Bootstrap tree #2, logLikelihood: -3331.565943\n",
      "[00:00:08] [worker #0] Bootstrap tree #1, logLikelihood: -2980.821417\n",
      "[00:00:11] [worker #3] Bootstrap tree #4, logLikelihood: -3462.602577\n",
      "[00:00:11] [worker #7] Bootstrap tree #8, logLikelihood: -3205.582452\n",
      "[00:00:16] [worker #1] Bootstrap tree #10, logLikelihood: -2886.183952\n",
      "[00:00:16] [worker #5] Bootstrap tree #14, logLikelihood: -3061.717206\n",
      "[00:00:16] [worker #0] Bootstrap tree #9, logLikelihood: -3238.111164\n",
      "[00:00:17] [worker #2] Bootstrap tree #11, logLikelihood: -3119.094118\n",
      "[00:00:19] [worker #3] Bootstrap tree #12, logLikelihood: -3236.218405\n",
      "[00:00:20] [worker #7] Bootstrap tree #16, logLikelihood: -3327.467091\n",
      "[00:00:21] [worker #4] Bootstrap tree #13, logLikelihood: -3197.206340\n",
      "[00:00:23] [worker #6] Bootstrap tree #15, logLikelihood: -3427.691748\n",
      "[00:00:24] [worker #5] Bootstrap tree #22, logLikelihood: -3125.609053\n",
      "[00:00:26] [worker #1] Bootstrap tree #18, logLikelihood: -3125.574360\n",
      "[00:00:27] [worker #2] Bootstrap tree #19, logLikelihood: -3393.479213\n",
      "[00:00:27] [worker #3] Bootstrap tree #20, logLikelihood: -3358.207761\n",
      "[00:00:27] [worker #0] Bootstrap tree #17, logLikelihood: -3296.773629\n",
      "[00:00:32] [worker #7] Bootstrap tree #24, logLikelihood: -3274.724357\n",
      "[00:00:33] [worker #6] Bootstrap tree #23, logLikelihood: -3299.690069\n",
      "[00:00:33] [worker #4] Bootstrap tree #21, logLikelihood: -3227.800483\n",
      "[00:00:34] [worker #5] Bootstrap tree #30, logLikelihood: -3296.257742\n",
      "[00:00:36] [worker #2] Bootstrap tree #27, logLikelihood: -3014.622690\n",
      "[00:00:36] [worker #0] Bootstrap tree #25, logLikelihood: -3090.048531\n",
      "[00:00:36] [worker #1] Bootstrap tree #26, logLikelihood: -3156.301071\n",
      "[00:00:36] [worker #3] Bootstrap tree #28, logLikelihood: -3254.445414\n",
      "[00:00:40] [worker #7] Bootstrap tree #32, logLikelihood: -3127.027642\n",
      "[00:00:42] [worker #5] Bootstrap tree #38, logLikelihood: -3266.005817\n",
      "[00:00:43] [worker #6] Bootstrap tree #31, logLikelihood: -3222.437012\n",
      "[00:00:44] [worker #0] Bootstrap tree #33, logLikelihood: -3097.892248\n",
      "[00:00:44] [worker #3] Bootstrap tree #36, logLikelihood: -3046.084407\n",
      "[00:00:45] [worker #1] Bootstrap tree #34, logLikelihood: -3112.474262\n",
      "[00:00:45] [worker #2] Bootstrap tree #35, logLikelihood: -3272.683372\n",
      "[00:00:47] [worker #4] Bootstrap tree #29, logLikelihood: -3307.582962\n",
      "[00:00:50] [worker #7] Bootstrap tree #40, logLikelihood: -3141.307825\n",
      "[00:00:52] [worker #5] Bootstrap tree #46, logLikelihood: -3262.357963\n",
      "[00:00:53] [worker #3] Bootstrap tree #44, logLikelihood: -3287.576897\n",
      "[00:00:53] [worker #2] Bootstrap tree #43, logLikelihood: -3328.382266\n",
      "[00:00:53] [worker #1] Bootstrap tree #42, logLikelihood: -3021.752966\n",
      "[00:00:54] [worker #0] Bootstrap tree #41, logLikelihood: -2995.991396\n",
      "[00:00:57] [worker #6] Bootstrap tree #39, logLikelihood: -3459.907940\n",
      "[00:00:59] [worker #4] Bootstrap tree #37, logLikelihood: -2964.774449\n",
      "[00:01:00] [worker #7] Bootstrap tree #48, logLikelihood: -3243.864399\n",
      "[00:01:01] [worker #1] Bootstrap tree #50, logLikelihood: -3125.307470\n",
      "[00:01:03] [worker #0] Bootstrap tree #49, logLikelihood: -3205.517622\n",
      "[00:01:11] [worker #4] Bootstrap tree #45, logLikelihood: -3236.624602\n",
      "[00:01:13] [worker #6] Bootstrap tree #47, logLikelihood: -3213.248480\n",
      "[00:01:22] [worker #4] Bootstrap tree #53, logLikelihood: -3020.401655\n",
      "[00:01:22] [worker #5] Bootstrap tree #54, logLikelihood: -3236.856368\n",
      "[00:01:22] [worker #1] Bootstrap tree #58, logLikelihood: -3299.312014\n",
      "[00:01:23] [worker #2] Bootstrap tree #51, logLikelihood: -3207.202821\n",
      "[00:01:23] [worker #0] Bootstrap tree #57, logLikelihood: -3125.785280\n",
      "[00:01:23] [worker #7] Bootstrap tree #56, logLikelihood: -3190.087914\n",
      "[00:01:26] [worker #3] Bootstrap tree #52, logLikelihood: -3105.908186\n",
      "[00:01:33] [worker #7] Bootstrap tree #64, logLikelihood: -3116.884544\n",
      "[00:01:33] [worker #6] Bootstrap tree #55, logLikelihood: -3162.762451\n",
      "[00:01:34] [worker #0] Bootstrap tree #65, logLikelihood: -2899.275701\n",
      "[00:01:34] [worker #4] Bootstrap tree #61, logLikelihood: -3050.336354\n",
      "[00:01:34] [worker #1] Bootstrap tree #66, logLikelihood: -3115.619810\n",
      "[00:01:37] [worker #2] Bootstrap tree #59, logLikelihood: -3332.084487\n",
      "[00:01:39] [worker #5] Bootstrap tree #62, logLikelihood: -3270.563915\n",
      "[00:01:40] [worker #3] Bootstrap tree #60, logLikelihood: -3454.966853\n",
      "[00:01:42] [worker #7] Bootstrap tree #72, logLikelihood: -3013.970953\n",
      "[00:01:43] [worker #0] Bootstrap tree #73, logLikelihood: -3174.049177\n",
      "[00:01:45] [worker #6] Bootstrap tree #63, logLikelihood: -2871.759210\n",
      "[00:01:46] [worker #1] Bootstrap tree #74, logLikelihood: -3218.960250\n",
      "[00:01:47] [worker #2] Bootstrap tree #67, logLikelihood: -3135.968200\n",
      "[00:01:50] [worker #4] Bootstrap tree #69, logLikelihood: -3016.184823\n",
      "[00:01:51] [worker #5] Bootstrap tree #70, logLikelihood: -3186.757790\n",
      "[00:01:51] [worker #3] Bootstrap tree #68, logLikelihood: -3062.298052\n",
      "[00:01:52] [worker #0] Bootstrap tree #81, logLikelihood: -3159.251825\n",
      "[00:01:52] [worker #7] Bootstrap tree #80, logLikelihood: -3022.158942\n",
      "[00:01:56] [worker #6] Bootstrap tree #71, logLikelihood: -3063.281681\n",
      "[00:01:57] [worker #1] Bootstrap tree #82, logLikelihood: -3305.156257\n",
      "[00:01:58] [worker #2] Bootstrap tree #75, logLikelihood: -3040.992016\n",
      "[00:01:59] [worker #4] Bootstrap tree #77, logLikelihood: -3157.717910\n",
      "[00:02:00] [worker #3] Bootstrap tree #76, logLikelihood: -3138.775209\n",
      "[00:02:01] [worker #5] Bootstrap tree #78, logLikelihood: -3023.922308\n",
      "[00:02:03] [worker #7] Bootstrap tree #88, logLikelihood: -3147.845063\n",
      "[00:02:03] [worker #0] Bootstrap tree #89, logLikelihood: -3155.349316\n",
      "[00:02:06] [worker #6] Bootstrap tree #79, logLikelihood: -3217.398949\n",
      "[00:02:07] [worker #4] Bootstrap tree #85, logLikelihood: -2901.623908\n",
      "[00:02:07] [worker #2] Bootstrap tree #83, logLikelihood: -3143.800778\n",
      "[00:02:11] [worker #3] Bootstrap tree #84, logLikelihood: -3017.360713\n",
      "[00:02:11] [worker #7] Bootstrap tree #96, logLikelihood: -2980.357449\n",
      "[00:02:12] [worker #5] Bootstrap tree #86, logLikelihood: -3217.054681\n",
      "[00:02:14] [worker #1] Bootstrap tree #90, logLikelihood: -3309.664057\n",
      "[00:02:17] [worker #4] Bootstrap tree #93, logLikelihood: -2893.840706\n",
      "[00:02:18] [worker #2] Bootstrap tree #91, logLikelihood: -3105.357674\n",
      "[00:02:21] [worker #6] Bootstrap tree #87, logLikelihood: -3155.060604\n",
      "[00:02:22] [worker #5] Bootstrap tree #94, logLikelihood: -3238.567214\n",
      "[00:02:22] [worker #3] Bootstrap tree #92, logLikelihood: -3208.189527\n",
      "[00:02:24] [worker #0] Bootstrap tree #97, logLikelihood: -3043.685816\n",
      "[00:02:25] [worker #1] Bootstrap tree #98, logLikelihood: -3087.049087\n",
      "[00:02:26] [worker #2] Bootstrap tree #99, logLikelihood: -3307.724284\n",
      "[00:02:29] [worker #6] Bootstrap tree #95, logLikelihood: -3047.679392\n",
      "[00:02:30] [worker #3] Bootstrap tree #100, logLikelihood: -3213.165432\n",
      "\n",
      "Bootstrap trees saved to: /Users/lachlanblack/Documents/GitHub/EspY3-protein-evolution/data/cleaned/raxml/test/01_50id_bootstrap/mytree_bs.raxml.bootstraps\n",
      "\n",
      "Execution log saved to: /Users/lachlanblack/Documents/GitHub/EspY3-protein-evolution/data/cleaned/raxml/test/01_50id_bootstrap/mytree_bs.raxml.log\n",
      "\n",
      "Analysis started: 03-Nov-2025 18:57:12 / finished: 03-Nov-2025 18:59:43\n",
      "\n",
      "Elapsed time: 150.855 seconds\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Generate bootstraps\n",
    "\n",
    "# Output: {PREFIX}_bs.raxml.bootstraps\n",
    "run(f\"raxml-ng --bootstrap --msa {MSA} --model {MODEL} --bs-trees {BS_TREES} --prefix {OUTDIR / (PREFIX + '_bs')} --threads {THREADS} --redo\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "97097faf-cd81-44c9-a41d-e2ce4eec4a67",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "$ raxml-ng --support --tree ../../data/cleaned/raxml/test/01_50id_bootstrap/mytree.raxml.bestTree --bs-trees ../../data/cleaned/raxml/test/01_50id_bootstrap/mytree_bs.raxml.bootstraps --prefix ../../data/cleaned/raxml/test/01_50id_bootstrap/mytree_support --redo\n",
      "\n",
      "\n",
      "RAxML-NG v. 1.2.2 released on 30.04.2024 by The Exelixis Lab.\n",
      "Developed by: Alexey M. Kozlov and Alexandros Stamatakis.\n",
      "Contributors: Diego Darriba, Tomas Flouri, Benoit Morel, Sarah Lutteropp, Ben Bettisworth, Julia Haag, Anastasis Togkousidis.\n",
      "Latest version: https://github.com/amkozlov/raxml-ng\n",
      "Questions/problems/suggestions? Please visit: https://groups.google.com/forum/#!forum/raxml\n",
      "\n",
      "System: Apple M2, 8 cores, 8 GB RAM\n",
      "\n",
      "RAxML-NG was called at 03-Nov-2025 19:00:33 as follows:\n",
      "\n",
      "raxml-ng --support --tree ../../data/cleaned/raxml/test/01_50id_bootstrap/mytree.raxml.bestTree --bs-trees ../../data/cleaned/raxml/test/01_50id_bootstrap/mytree_bs.raxml.bootstraps --prefix ../../data/cleaned/raxml/test/01_50id_bootstrap/mytree_support --redo\n",
      "\n",
      "Analysis options:\n",
      "  run mode: Compute bipartition support (Felsenstein Bootstrap)\n",
      "  start tree(s): user\n",
      "  random seed: 1762196433\n",
      "  SIMD kernels: SSE3\n",
      "  parallelization: coarse-grained (auto), PTHREADS (auto)\n",
      "\n",
      "WARNING: Running in REDO mode: existing checkpoints are ignored, and all result files will be overwritten!\n",
      "\n",
      "Reading reference tree from file: ../../data/cleaned/raxml/test/01_50id_bootstrap/mytree.raxml.bestTree\n",
      "Reference tree size: 44\n",
      "\n",
      "Reading bootstrap trees from file: ../../data/cleaned/raxml/test/01_50id_bootstrap/mytree_bs.raxml.bootstraps\n",
      "Loaded 100 trees with 44 taxa.\n",
      "\n",
      "Best ML tree with Felsenstein bootstrap (FBP) support values saved to: /Users/lachlanblack/Documents/GitHub/EspY3-protein-evolution/data/cleaned/raxml/test/01_50id_bootstrap/mytree_support.raxml.support\n",
      "\n",
      "Execution log saved to: /Users/lachlanblack/Documents/GitHub/EspY3-protein-evolution/data/cleaned/raxml/test/01_50id_bootstrap/mytree_support.raxml.log\n",
      "\n",
      "Analysis started: 03-Nov-2025 19:00:33 / finished: 03-Nov-2025 19:00:33\n",
      "\n",
      "Elapsed time: 0.018 seconds\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Map boostrap support onto ML tree\n",
    "\n",
    "# Output: {PREFIX}_support.raxml.support -> final tree with bootstrap values\n",
    "run(f\"raxml-ng --support --tree {OUTDIR / (PREFIX + '.raxml.bestTree')} --bs-trees {OUTDIR / (PREFIX + '_bs.raxml.bootstraps')} --prefix {OUTDIR / (PREFIX + '_support')} --redo\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0bb3f27-9352-4a6c-84b0-532a5ebca514",
   "metadata": {},
   "source": [
    "# Visualising the tree\n",
    "\n",
    "Open '{PREFIX}_support.raxml.support' in FigTree\n",
    "- Show Node Labels -> label to see boostrap support values (0-100)\n",
    "- The scale bar shows substitutions per site\n",
    "- You can midpoint-root or outgroup-root for directionality"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
