{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8880d8a0-fa89-4304-886d-8f3bd249063d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "d22aa90d-d33a-42b7-8560-8936f14ab6b8",
   "metadata": {},
   "source": [
    "# EspY3 flanking-region homologue search\n",
    "## 4/11/25\n",
    "\n",
    "EspY3 has a domain of two pentapeptide repeats (PPRs). This is the only annotated region in Pfam and is conserved across 10,000 proteins. Whilst it is highly conserved and structurally stable, the surrounding flanks evolve more rapidly and often determine function or interactions. Variation in these flanks therefore can signal functional divergence or specialisation among homologous proteins. Phylogenetic analysis of these regions may uncover lineage-specific adaptations and link sequence evolution outside the repeat domain to potential differences in protein function within pathogens. \n",
    "\n",
    "\n",
    "In this notebook I aim to:\n",
    "1. Extract the N- and C-terminal flanks *outside* the PPR domain (165-342.\n",
    "2. Search BLASTP with each flanking sequence to find homologues\n",
    "3. Curate hits with CD-HIT to remove redundancy\n",
    "4. Annotate with Pfam/InterPro to see if any new domains appear in the flanks\n",
    "5. Build allignments and trees\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "276219ce-f1da-4db0-b5c1-4e76658cc262",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[INFO]\u001b[0m create or read FASTA index ...\n",
      "[INFO]\u001b[0m read FASTA index from ../../data/raw/espy3.fasta.seqkit.fai\n",
      "[INFO]\u001b[0m   1 records loaded from ../../data/raw/espy3.fasta.seqkit.fai\n",
      "[INFO]\u001b[0m create or read FASTA index ...\n",
      "[INFO]\u001b[0m read FASTA index from ../../data/raw/espy3.fasta.seqkit.fai\n",
      "[INFO]\u001b[0m   1 records loaded from ../../data/raw/espy3.fasta.seqkit.fai\n"
     ]
    }
   ],
   "source": [
    "# 1. Isolating the N- and C- flanks.\n",
    "\n",
    "# N flank residues 1 - 164\n",
    "!seqkit subseq -r 1:164 ../../data/raw/espy3.fasta > ../../results/flank-search/flank_N.fasta\n",
    "\n",
    "# C flank residues 343 - 523\n",
    "!seqkit subseq -r 343:-1 ../../data/raw/espy3.fasta > ../../results/flank-search/flank_C.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4797b66-4d48-4813-9134-86c267ab819c",
   "metadata": {},
   "source": [
    "## 2. Blast search\n",
    "\n",
    "I used the online NCBI BLAST. I searched each flanking region against the refseq_protein database, and set the max target sequences to 5000. All other settings are default blastp.\n",
    "\n",
    "The output of both flanks shows that the flanking regions are restricted to enterobacterales. This supports a model where this protein evolved from a general PPR scaffold but specialised within enteric pathogens."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
